File: schemas/video-conferencing/drizzle/meeting_invitations.ts
// meeting_invitations: Invitations sent to users for upcoming meetings.
// See README.md for full design rationale.

import { pgTable, pgEnum, uuid, text, timestamp, index, unique } from "drizzle-orm/pg-core";
import { meetings } from "./meetings";
import { users } from "./users";

export const invitationStatusEnum = pgEnum("invitation_status", ["pending", "accepted", "declined", "tentative"]);

export const meetingInvitations = pgTable(
  "meeting_invitations",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    meetingId: uuid("meeting_id").notNull().references(() => meetings.id, { onDelete: "cascade" }),
    inviterId: uuid("inviter_id").notNull().references(() => users.id, { onDelete: "cascade" }),
    inviteeId: uuid("invitee_id").notNull().references(() => users.id, { onDelete: "cascade" }),
    status: invitationStatusEnum("status").notNull().default("pending"),
    respondedAt: timestamp("responded_at", { withTimezone: true }),
    message: text("message"),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    unique("uq_meeting_invitations_meeting_id_invitee_id").on(table.meetingId, table.inviteeId),
    index("idx_meeting_invitations_invitee_id_status").on(table.inviteeId, table.status),
    index("idx_meeting_invitations_meeting_id_status").on(table.meetingId, table.status),
  ]
);

File: schemas/video-conferencing/spacetimedb/meeting_invitations.rs
// meeting_invitations: invitations sent to users for upcoming meetings.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum InvitationStatus {
    Pending, // type: String
    Accepted,
    Declined,
    Tentative,
}

// Composite unique (meeting_id, invitee_id) — enforce in reducer logic

#[spacetimedb::table(name = meeting_invitations, public)]
pub struct MeetingInvitation {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub meeting_id: String, // UUID — FK → meetings.id (cascade delete)
    pub inviter_id: String, // UUID — FK → users.id (cascade delete)
    #[index(btree)]
    pub invitee_id: String, // UUID — FK → users.id (cascade delete)
    pub invitation_status: InvitationStatus,
    pub responded_at: Option<Timestamp>,
    pub message: Option<String>,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
